### DISCUSSION TITLE
Where do global project configs go in a multi-module project?

### DISCUSSION THREAD

**Discussion by JordanLongstaff** [kotest/kotest#4361] [Q&A]
I've got a multi-module project with lots of submodules using Kotest. I'd like to configure a couple of properties (e.g. parallelism) globally across all of the modules, by having the configurations in **one place**, and I'm currently at a loss.

I tried putting the configuration in the top-level `gradle.properties` file, and that didn't work (it seems Kotest just doesn't look there, if I'm understanding the docs correctly).

I tried putting it in the `test/resources` folder of one of the modules, which is the parent of all the other submodules that use Kotest. However, it only worked for the parent module.

I tried putting it in a submodule instead, because most of these modules, including the parent module, depend on each other, so I picked one that doesn't depend on any others, hoping the configs would propagate through dependencies. They did not.

Of course I've tried an `AbstractProjectConfig` object too, but it has all the same problems as system properties.

Is it just a case of not knowing where to put it, or do I have no choice but to duplicate the configs in every module?

[Here](https://github.com/JordanLongstaff/ArtemisAgent/tree/main/IAN) is a link to the aforementioned parent module in my project, in case the architecture yields any ideas of where one might put a global config, if it's feasible.

**Comment by Kantis** [comment]
@JordanLongstaff seeing as you already have a `testing` module which your test sources depend upon (and where your `ProjectConfig` resides), it's quite simple.

I would suggest just adding your `kotest.properties` file in `src/main/resources` in the testing module. That would remove the need to duplicate it across all modules, as long as you include the `testImplementation(projects.ian.testing)` dependency.

PS. If you add `kotest.framework.classpath.scanning.autoscan.disable=true` you should also see quicker test start times.

**Comment by sksamuel** [comment]
As @Kantis says, a kotest.properties file the top level module on the classpath would be picked up by all tests.

**Reply by JordanLongstaff** [reply]
OK, yes, I just added a common `kotest.properties` to the root folder of my project, and it seems to work. (The one thing I hadn't tried, LOL)

**Reply by JordanLongstaff** [reply]
Never mind. It does not work at all. So it's back to the setup I had before.

**Comment by JordanLongstaff** [comment]
I don't like the idea of making a module just for a shared test config. I only did it because #4643 didn't exist, and it's causing a lot of false positives with a [dependency analysis plugin](https://github.com/autonomousapps/dependency-analysis-gradle-plugin). As soon as there's an upgrade to Kotest that includes #4643 (and #4340), I'm going to remove that module.

I'll do separate `kotest.properties` for different modules if I have to. They're already varied in the parallelism setting.

**Reply by sksamuel** [reply]
Aiming to get an M2 build out as soon as I can with the latest round of changes.

**Reply by JordanLongstaff** [reply]
> I'll do separate `kotest.properties` for different modules if I have to. They're already varied in the parallelism setting.

If I had a shared `kotest.properties` file that included everything but the parallelism setting, and then each module has their own file that defines it, would the properties merge?

**Reply by sksamuel** [reply]
There are tons of changes around shared config in the latest main build, and hoping to cut M2 today, which you could try.

**Comment by sksamuel** [comment]
Closing this as we now support package level configs
